core/src/main/java/org/apache/iceberg/RemoveSnapshots.java

```diff
@@ -163,21 +188,100 @@ private TableMetadata internalApply() {
+    }
+
+    idsToRetain.addAll(idsOfRetainedRefs);
+
+    List<Map.Entry<String, SnapshotRef>> branches = base.refs().entrySet().stream()
+        .filter(refEntry -> refEntry.getValue().isBranch())
+        .collect(Collectors.toList());
+
+    Set<Long> branchSnapshotsToRetain = computeAllBranchSnapshotsToRetain(branches);
+    idsToRetain.addAll(branchSnapshotsToRetain);
+    TableMetadata.Builder updatedMetaBuilder = TableMetadata.buildFrom(base);
+    idsToRemove.addAll(base.snapshots().stream().map(Snapshot::snapshotId)
+        .filter(snapshot -> !idsToRetain.contains(snapshot)).collect(Collectors.toSet()));
+    updatedMetaBuilder.removeSnapshots(idsToRemove);
+    expiredRefs.forEach(updatedMetaBuilder::removeRef);
+    TableMetadata updatedMetadata = updatedMetaBuilder.build();
+    List<Snapshot> updateSnapshots = updatedMetadata.snapshots();
+    List<Snapshot> baseSnapshots = base.snapshots();
+    boolean updated = updateSnapshots.size() != baseSnapshots.size() || expiredRefs.size() > 0;
+    return updated ? updatedMetadata : base;
+  }
+
+  private Set<Long> computeAllBranchSnapshotsToRetain(List<Map.Entry<String, SnapshotRef>> branches) {
+    Set<Long> branchSnapshotsToRetain = Sets.newHashSet();
+    for (Map.Entry<String, SnapshotRef> branchEntry : branches) {
+      String name = branchEntry.getKey();
+      SnapshotRef branch = branchEntry.getValue();
+      Long expireSnapshotsOlderThan;
+      Integer minSnapshotsToKeep;
+      if (name.equals(SnapshotRef.MAIN_BRANCH) || branch.maxSnapshotAgeMs() == null) {
+        expireSnapshotsOlderThan = this.expireOlderThan;
+      } else {
+        expireSnapshotsOlderThan = now - branch.maxSnapshotAgeMs();
+      }
+
+      if (name.equals(SnapshotRef.MAIN_BRANCH) || branch.minSnapshotsToKeep() == null) {
+        minSnapshotsToKeep = this.minNumSnapshots;
+      } else {
+        minSnapshotsToKeep = branch.minSnapshotsToKeep();
+      }
+
+      branchSnapshotsToRetain.addAll(
+          computeBranchSnapshotsToRetain(branch, expireSnapshotsOlderThan, minSnapshotsToKeep));
+    }
+    return branchSnapshotsToRetain;
+  }
+
+  private Set<String> computeExpiredRefs(Map<String, SnapshotRef> refs) {
+    Set<String> expiredRefs = Sets.newHashSet();
+    for (String name : refs.keySet()) {
+      SnapshotRef ref = refs.get(name);
+      Long maxRefAgeMs = ref.maxRefAgeMs() != null ? ref.maxRefAgeMs() : this.defaultMaxRefAgeMs;
+      if (!name.equals(SnapshotRef.MAIN_BRANCH)) {
+        Snapshot snapshot = base.snapshot(ref.snapshotId());
+        if (now - snapshot.timestampMillis() >= maxRefAgeMs) {
+          expiredRefs.add(name);
+        }
+      }
+    }
+    return expiredRefs;
+  }
+
+  private Set<Long> computeBranchSnapshotsToRetain(SnapshotRef branch, Long expireSnapshotsOlderThan,
+      Integer minSnapshotsToKeep) {
+    Snapshot snapshot = base.snapshot(branch.snapshotId());
+    List<Long> ancestorIds = SnapshotUtil.ancestorIds(snapshot, base::snapshot);
```

Comment:
You might consider doing this a bit differently. I think it would be more direct to use the Iterable:

```java
  Set<Long> idsToRetain = Sets.newHashSet();
  for (Snapshot snap : SnapshotUtil.ancestorsOf(snapshotId, base::snapshot)) {
    if (idsToRetain.size() < minSnapshotsToKeep || snap.timestampMillis() >= expireOlderThan) {
      idsToRetain.add(snap.snapshotId());
    } else {
      break;
    }
  }

  return idsToRetain;
```

Reply:
Definitley, the iterable approach is much cleaner. Updated